lock/src/main/java/org/infinispan/lock/impl/lock/ClusteredLockImpl.java

```diff
@@ -221,34 +255,95 @@ public void entryRemoved(CacheEntryRemovedEvent event) {
-                           }));
-                  }
-               } catch (AvailabilityException ex) {
-                  log.error("Unable to release due to cluster change", ex);
-               }
-            });
+         Set<Object> leavingNodes = oldMembers.stream().filter(a -> !newMembers.contains(a)).collect(Collectors.toSet());
+
+         if (leavingNodes.isEmpty()) {
+            log.tracef("LOCK[%s] Nothing to do, all nodes are present %s", getName(), originator);
+            return;
+         }
+
+         if (leavingNodes.size() >= newMembers.size() && oldMembers.size() > 2) {
+            // If the oldMembers size is greater than 2, we do nothing because the other nodes will handle
+            // If the cluster was formed by 2 members and one leaves, we should not enter here
+            log.tracef("LOCK[%s] Nothing to do, we are on a minority partition notification on %s", getName(), originator);
+            return;
+         }
+
+         if (clusteredLockManager.isDefined(name)) {
+            log.tracef("LOCK[%s] %s launches unlock for each leaving node", getName(), originator);
+            forceUnlockForLeavingMembers(leavingNodes);
          }
       }
+
+      /**
+       * This method forces unlock for each of the Address that is not present in the cluster. We don't know which node
+       * holds the lock, so we force an unlock
+       *
+       * @param possibleOwners
+       */
+      private void forceUnlockForLeavingMembers(Set<Object> possibleOwners) {
+         clusteredLockManager.execute(() -> {
+                  log.tracef("LOCK[%s] Call force unlock for %s from %s ", getName(), possibleOwners, originator);
+                  int viewChangeUnlockValue = viewChangeUnlockHappening.incrementAndGet();
+                  log.tracef("LOCK[%s] viewChangeUnlockHappening value in %s ", getName(), viewChangeUnlockValue, originator);
+                  unlock(null, possibleOwners)
+                        .whenComplete((unlockResult, ex) -> {
+                           log.tracef("LOCK[%s] Force unlock call completed for %s from %s ", getName(), possibleOwners, originator);
+                           int viewChangeUnlockValueAfterUnlock = viewChangeUnlockHappening.decrementAndGet();
+                           log.tracef("LOCK[%s] viewChangeUnlockHappening value in %s ", getName(), viewChangeUnlockValueAfterUnlock, originator);
+                           if (ex == null) {
+                              log.tracef("LOCK[%s] Force unlock result %b for %s from %s ", getName(), unlockResult, possibleOwners, originator);
+                           } else {
+                              log.error(ex, log.unlockFailed(getName(), getOriginator()));
+                              // TODO: handle the exception. Retry ? End all the pending requests in this lock ?
+                           }
+
+                           retryPendingUnlocks();
+                           retryPendingRequests();
```

Comment:
I understand unlocking already acquires the lock for the new requestor, so is this really necessary?

Reply:
retryPendingRequests oui, unlocks no

Reply:
Doesn't `retryPendingUnlocks()` unlocking the locks trigger `LockReleasedListener`? Or is the problem that `LockReleasedListener` isn't notified reliably?